### ISSUE TITLE
Custom token generation as an option

### ISSUE THREAD

**Issue by maletor** [simplabs/rails_api_auth#28]
Unless there is something totally flawed with my logic, the last step of generating a new, separate access token is unnecessary where login is via OAuth.

It would be more powerful to forward the access tokens exactly as acquired back to the client. For a few reasons.

1. Less tokens. We don't expect this app to have its own other client apps. A lean Rails API would have a separate resource server and authentication server. By sending a quick "verify token" request to Facebook we can assure its validity even if Facebook has revoked the user. Further, you could use JWT and do token verification with an air gap if you owned the authentication server and resource server.

2. Access third party APIs directly from the client. Ember, for instance, beautifully sets up adapters for models. So consider we had a user model in our SPA and authenticated with a Facebook access token. We can make requests directly to the Graph API host for user info that never goes stale. Think about the power with an access token belonging to a large suite of shared authorization microservices. Sideloading disparate objects becomes a non-issue so long as CORS is properly configured.

3. Users not wanting to authenticate via OAuth with your client API can use Basic Auth at Facebook to get a token directly so that they may start using your API. Of course, the client API is still around to service requests for token generation but that's mostly a convenience. Either token generation method should be valid.

You might say that exposing the Facebook access token versus the client API's access token is less secure. In a worse case scenario situation where you have an XSS vulnerability the attack surface is larger by the difference between what your app does with the scopes granted and the total scopes granted. Again, if you owned the authentication and resource server this can be reduced to something really small.

**Comment by marcoow** [comment]
The idea of this library is really more that you'd use Facebook, Google etc. as a means of validating the user's identity not to expose their access tokens directly to the client. What you usually want to have is multiple ways of authenticating users when they use your application - be it against your server directly via username/password or be it an external provider like Facebook etc. In any case you want users to end up with a token that's been issued by your application and can be used to e.g. look up the respective user in the DB.

**Comment by maletor** [comment]
That idea isn't excluded by what I suggested. I can imagine a user would
send as many access tokens as the client is wanting to support. Maybe the
user could even pass a hint in a custom header to speed up the process of
validating it.

**Comment by marcoow** [comment]
Not sure what you mean. When you're using bearer token auth you can only send one bearer token in the `Authorization` header per request.

**Comment by marcoow** [comment]
@maletor: any update on this?

**Comment by maletor** [comment]
Yes, you can only send one, but that doesn't stop your server from validating it at more than one place.
